File: Util/parse_test.cc
#include "Util/parse.h"

#include <iostream>
#include <string>

#include "Util/assert.h"

using namespace cwerg;
using namespace std;

void EmitId(string_view s) { cout << "{" << s << "}"; }

void EmitString(string_view s) {
  if (s.size() == 2) {
    cout << s;
    return;
  }

  char buffer[4 * 1024];
  ASSERT(s.size() < sizeof(buffer), "string too large");

  uint32_t size = EscapedStringToBytes({s.data() + 1, s.size() - 2}, buffer);
  if (size == 0) {
    cout << "@FAILED@";
    return;
  }

  char buffer2[4 * 1024];
  size = BytesToEscapedString({buffer, size}, buffer2);
  buffer2[size] = 0;
  cout << "\"" << buffer2 << "\"";
}

std::vector<std::string_view> split(const std::string_view& s, char delim) {
  std::vector<std::string_view> out;
  int token_start = -1;
  for (int i = 0; i < s.size(); ++i) {
    char c = s[i];
    if (c == delim) {
      if (token_start >= 0) {
        out.push_back(s.substr(token_start, i - token_start));
        token_start = -1;
      }
    } else if (token_start < 0) {
      token_start = i;
    }
  }
  if (token_start >= 0) {
    out.push_back(s.substr(token_start, s.size() - token_start));
  }
  return out;
}

int main(int argc, char* argv[]) {
  std::string mode = argv[1];
  for (string line; getline(std::cin, line);) {
    cout << "\n" << line << "\n";
    std::vector<std::string_view> vec;
    if (mode == "lex") {
      if (!ParseLineWithStrings(line.c_str(), false, &vec)) {
        cout << "@FAILED@\n\n";
        continue;
      }
      const char* sep = "";
      for (const string_view& s : vec) {
        cout << sep;
        sep = " ";
        ASSERT(s.size() > 0, "");
        if (s[0] == '"') {
          EmitString(s);
        } else {
          EmitId(s);
        }
      }
      cout << "\n";
    } else {
      ASSERT(mode == "num", "");
      std::vector<std::string_view> vec = split(line, ' ');
      if (vec.size() == 2) {
        char buf[32];
        if (vec[0] == "int64") {
          cout << "[INT64] ";
          auto val = ParseInt<int64_t>(vec[1]);
          if (!val.has_value()) {
            cout << "@BAD VALUE@\n";
            continue;
          }
          cout << ToDecSignedString(val.value(), buf) << " "
               << ToHexString(val.value(), buf) << "\n";
        } else if (vec[0] == "uint64") {
          cout << "[UINT64] ";
          auto val = ParseInt<uint64_t>(vec[1]);
          if (!val.has_value()) {
            cout << "@BAD VALUE@\n";
            continue;
          }
          cout << ToDecString(val.value(), buf) << " "
               << ToHexString(val.value(), buf) << "\n";
        } else if (vec[0] == "flt64") {
          cout << "[FLT64] ";
          auto val = ParseFlt64(vec[1]);
          if (!val.has_value()) {
            cout << "@BAD VALUE@\n";
            continue;
          }
          cout << ToFltString(val.value(), buf) << " "
               << ToFltHexString(val.value(), buf) << "\n";
        } else if (vec[0] == "char") {
          cout << "[CHAR] ";
          auto val = ParseChar(vec[1]);
          if (!val.has_value()) {
            cout << "@BAD VALUE@\n";
            continue;
          }
          cout << val.value() << "\n";
        }
      }
    }
  }
  return 0;
}

File: Util/parse_test.py
#!/bin/env python3

"""

"""
from typing import List
import sys
import struct

from Util import parse


def EmitString(s: str):
    b = parse.EscapedStringToBytes(t[1:-1])
    t2 = parse.BytesToEscapedString(b)
    print('"' + t2 + '"', end="")


def EmitId(s: str):
    print("{" + t + "}", end="")


def Flt64FromBits(data: int) -> float:
    return struct.unpack('<d', int.to_bytes(data, 8, "little"))[0]


def Flt64ToBits(num: float) -> int:
    b = struct.pack('<d', num)
    assert len(b) == 8
    return int.from_bytes(b, "little")


if __name__ == '__main__':
    mode = sys.argv[1]
    for line in sys.stdin:
        print()
        line = line[:-1]
        print(line)
        if mode == "lex":
            try:
                token: List[str] = parse.ParseLine(line)
                # print (repr(token))
                sep = ""
                for t in token:
                    print(sep, end="")
                    sep = " "
                    if t.startswith('"'):
                        EmitString(t)
                    else:
                        EmitId(t)

            except Exception as err:
                print("@FAILED@")
            print()
        else:
            assert mode == "num"
            token = line.split()
            assert len(token) == 2
            if token[0] == "uint64":
                val = parse.ParseUint64(token[1])
                if val is None:
                    print(f"[UINT64] @BAD VALUE@")
                else:
                    print(f"[UINT64] {val} {val:x}")
            elif token[0] == "int64":
                val = parse.ParseInt64(token[1])
                if val is None:
                    print(f"[INT64] @BAD VALUE@")
                else:
                    print(f"[INT64] {val} {val & ((1 << 64) - 1):x}")
            elif token[0] == "flt64":
                if token[1].startswith("0x"):
                    out = int(token[1], 0)
                    val = Flt64FromBits(out)
                else:
                    val = float(token[1])
                if val is None:
                    print(f"[FLT64] @BAD VALUE@")
                else:
                    print(f"[FLT64] {val:g} {Flt64ToBits(val):x}")
            elif token[0] == "char":
                val = parse.ParseChar(token[1])
                if val is None:
                    print(f"[CHAR] @BAD VALUE@")
                else:
                    print(f"[CHAR] {val}")
